.github/workflows/tests.yml

```diff
@@ -73,39 +73,45 @@ jobs:
         | jq '.files."time_stats.csv".raw_url' \
         | xargs curl > .tmp/time_stats.csv
 
-    - name: Upload shared artifacts
-      uses: actions/upload-artifact@v2
-      with:
-        name: build-cache-key-and-time-stats
-        path: .tmp
-        retention-days: 1
-
     - name: Install system deps
       if: steps.buildcache.outputs.cache-hit != 'true'
       run: |
         sudo apt-get update
         sudo apt-get install -y uuid-dev libreadline-dev bison flex
 
     - name: Install rust toolchain
-      if: steps.buildcache.outputs.cache-hit != 'true'
+      if: steps.buildcache.outputs.cache-hit != 'true' || steps.venvcache.outputs.cache-hit != 'true'
       uses: actions-rs/toolchain@v1
       with:
         profile: minimal
         toolchain: stable
         default: true
 
     - name: Build
-      if: steps.buildcache.outputs.cache-hit != 'true'
+      if: steps.buildcache.outputs.cache-hit != 'true' || steps.venvcache.outputs.cache-hit != 'true'
       run: |
         # --no-use-pep517 because we have explicitly installed all deps
         # and don't want them to be reinstalled in an "isolated env".
         pip install --no-use-pep517 --no-deps -e .[test,docs]
 
     - name: Bootstrap EdgeDB Server
-      if: steps.buildcache.outputs.cache-hit != 'true'
+      if: steps.buildcache.outputs.cache-hit != 'true' || steps.venvcache.outputs.cache-hit != 'true'
```

Comment:
If I'm reading this correctly, the build step and the bootstrap step share the same cache?  Will bootstrap invalidation affect the build cache too?  Many PRs change the core source and so bootstrap cache would be invalidated much more frequently than the build cache which depends only on deps and Rust bits.

Reply:
> the build step and the bootstrap step share the same cache?

Yes.

> Will bootstrap invalidation affect the build cache too?

Yes, because they are placed in the same `build/` folder, though the impact is partial - Github Action would still be able to pull in the previous cache as a partial match (`cache-hit = false`) of `restore-keys`. Looks like `pip install -e .` will still re-build the extensions, but the time is shorter than a complete rebuild.

Are the pickle files the only outcome of the bootstrap, please?  I wasn't sure if the `build/` directory was also updated. If we could separate the bootstrap cache, that would be the best. Actually, if we could confirm the following, I could make the caching more fine-grained and efficient, we don't even need the `build/` folder in each shard (?):

* `setup.py` -> virtualenv
* `postgres` + `pgproto` + Rust/Cython Files -> `*.so`
* `CACHE_SRC_DIRS` -> `*.pickle`
* `edb/edgeql/parser/grammer/**/*.py` -> ???
* edgedb-cli -> `edb/cli/edgedb`

I'm sorry if this is a dumb question due to my ignorance, but I'd like to ask here first before I spend too much time reading the wrong part of the code.

Reply:
Never mind - I think I figured this out, will do some more testing here.

Reply:
Can we have a separate cache for every item here:

> virtualenv + postgres + pgproto + Rust/Cython Files -> *.so

?

E.g. "postgres" is very unlikely to change, so it should always be cached separately (building it takes a while).

Reply:
> Are the pickle files the only outcome of the bootstrap to be cached

The outcome of bootstrap is the contents of `build/cache/`, which includes a schema pickle and an SQL dump, so `CACHE_SRC_DIRS` -> `build/cache`

> `edb/edgeql/parser/grammer/**/*.py` -> 

these produce the parser table pickles (which are stored next to them). See `_compile_parsers` in `setup.py`.  All parsers share the grammar and so should be cached together (just take the hash of `edb.edgeql.parser.grammar` and `edgeql-parser/src/keywords.rs`.